node/cn/tracers/api.go

```diff
@@ -429,34 +429,12 @@ func (api *CommonAPI) traceChain(start, end *types.Block, config *TraceConfig, n
 				logged = time.Now()
 				logger.Info("Tracing chain segment", "start", start.NumberU64(), "end", end.NumberU64(), "current", number, "transactions", traced, "elapsed", time.Since(begin))
 			}
-			// Retrieve the parent block and target block for tracing.
-			block, err := api.blockByNumber(localctx, rpc.BlockNumber(number))
-			if err != nil {
-				failed = err
-				break
-			}
 			next, err := api.blockByNumber(localctx, rpc.BlockNumber(number+1))
 			if err != nil {
 				failed = err
 				break
 			}
-			// Prepare the statedb for tracing. Don't use the live database for
-			// tracing to avoid persisting state junks into the database. Switch
-			// over to `preferDisk` mode only if the memory usage exceeds the
-			// limit, the trie database will be reconstructed from scratch only
-			// if the relevant state is available in disk.
-			var preferDisk bool
-			if statedb != nil {
-				s1, s2, s3 := statedb.Database().TrieDB().Size()
-				preferDisk = s1+s2+s3 > defaultTracechainMemLimit
```

Comment:
@hyunsooda Can you revive this logic? arguments `StateAtBlock(..., base *state.StateDB, readOnly bool, preferDisk bool(` might have a huge impact on performance. I'm okay with adding the arguments to StateAtTransaction.

Reply:
Sorry, I don't understand what you are concerned about. I replaced `StateAtBlock` to `StateAtTransaction`. With the argument tx index zero, the loop cost would be zero. Could you explain more about this?

Reply:
In StateAtTransaction(block, txIndex, reexec), it calls `StateAtBlock(parent, reexec, nil, true, false)`. The last 3 arguments are fixed. But it seems TraceChain wants to customize those 3 arguments to be `base, false, preferDisk`. By replacing StateAtBlock with StateAtTransaction, the 3 arguments are now fixed to be `nil, true, false`. I think the performance might degrade by the change.

So I suggest you to allow StateAtTransaction to receive the 3 arguments and pass it to StateAtBlock.

Reply:
PTAL https://github.com/kaiachain/kaia/pull/121/commits/b886f3f020ac4ca2adbeb68308b6b40b54e0ac8a